spices/SPICE-0009-external-readers.adoc

```diff
@@ -0,0 +1,319 @@
+
+  /// Register client-side resource readers.
+  ///
+  /// [ClientResourceReader] is defined at https://pkl-lang.org/main/current/bindings-specification/message-passing-api.html#create-evaluator-request
+  clientResourceReaders: Listing<ClientResourceReader>?
+}
+----
+
+`CreateEvaluatorRequest` will be expanded with an additional property:
+[source,pkl]
+----
+externalReaders: Listing<ExternalReader>?
+
+class ExternalReader {
+  /// May be specified as an absolute path to an executable
+  /// May also be specified as just an executable name, in which case it will be resolved according to the PATH environment variable
+  executable: String
+
+  /// Command line arguments that will be passed to the reader process
+  arguments: Listing<String>
+}
+----
+
+=== CLI
+
+A new `--external-reader` CLI argument will be added to configure external readers.
+This argument can be provided multiple times (using `clikt.multiple`) to configure multiple external readers.
+The argument may be passed as a space-separated string where the first element becomes `executable` and any remainder becomes `arguments`.
+
+TBD: It might be best if the argument value is link:https://docs.python.org/3/library/shlex.html#shlex.split[shlex'd] instead of split to support passing arguments to the reader process that contain spaces.
+
+=== Standard Library
+
+The `EvaluatorSettings` module will be expanded to enable configuring external readers in `PklProject` files:
+
+[source,pkl]
+----
+externalReaders: Listing<ExternalReader>?
+
+class ExternalReader {
+  /// May be specified as an absolute path to an executable
+  /// May also be specified as just an executable name, in which case it will be resolved according to the PATH environment variable
+  executable: String
+
+  /// Command line arguments that will be passed to the reader process
+  arguments: Listing<String>
+}
+----
+
+=== Language Binding Libraries
+
+The language binding libraries `pkl-go` and `pkl-swift` will be expanded to support using and implementing external readers.
+For the purpose of illustration, examples will be provided using Golang.
+
+The `EvaluatorOptions` type will be expanded to include a new property for external readers:
+
+[source,go]
+----
+type EvaluatorOptions struct {
+	// ...
+    ExternalReaders []ExternalReader
+	// ...
+}
+
+type ExternalReader struct {
+	Executable string
+	Arguments []string
+}
+----
+
+A new `ExternalReaderRuntime` type will be introduced to implement the child process message passing interface.
+It makes sense to expand the existing libraries to add this functionality as much of the message passing infrastructure and types for implementing resource and module readers can be reused.
+An `ExternalReaderRuntime` is configured with zero or more `ResourceReader` instances and zero or more `ModuleReader`.
+When started, the runtime will consume messages from standard input, dispatch calls to the configured readers, and send responses over standard output.
```

Comment:
Probably better to be passed a reader/writer interface

Reply:
Great idea! Updated